### DISCUSSION TITLE
Mutation result missing data? Graphcache update may be at fault.

### DISCUSSION THREAD

**Discussion by orbiteleven** [urql-graphql/urql#2050] [Q&A] [answered]
In the off chance that someone has a similar problem I'm going to throw this out there:

I'm seeing cases where `data.<mutationName>` does not return data. I'm pretty sure this has to do with resolver cache updates, as it only occurs when I'm using them. However, I could be just screwing this up.

Given a simple schema

```graphql
mutation CreateItem(
  $name: String!
  $parentId: String!
) {
  createItem(
    data: {
      name: $name
      parentId: $parentId
    }
  ) {
    id
    name
    Parent {
      id
      name
    }
  }
}
```

and  helper function: 
```typescript
interface AddItemToFieldListParams<Target, Item> {
  cache: Cache
  target: Target
  field: string
  item: Item
  prepend?: boolean
}

export const addItemToFieldList = <TTarget, TItem>({
  cache,
  target,
  field,
  item,
  prepend = false,
}: AddItemToFieldListParams<TTarget, TItem>): void => {
  const Target = cache.keyOfEntity({ ...(target as WithTypeName<TTarget>) })
  const Item = cache.keyOfEntity({ ...(item as WithTypeName<TItem>) })
  cache.inspectFields(Target).forEach(({ fieldName, fieldKey }) => {
    if (fieldName === field) {
      const list = [
        ...[prepend ? Item : undefined],
        ...(cache.resolve(Target, fieldKey) as NullArray<string>),
        ...[prepend ? undefined : Item],
      ]
      cache.link(Target, fieldKey, list as NullArray<string | Data>)
    }
  })
}
```

and a cache update:

```typescript
export type ItemWithType = WithTypeName<Item>

interface ItemData {
  createItem?: ItemWithType
}

export const createItem: UpdateResolver<ItemData> = (
  result,
  _args,
  cache,
  _info
) => {
  // *** This is fine ***
  result.createItem.Followers.forEach((connection) => {
    addItemToFieldList({
      cache,
      target: connection.Follower,
      field: 'Following',
      item: connection,
    })
  })

  // *** This will clear `data.createItem` ***
  if (result.createItem.Parent) {
    const Parent = result.createItem.Parent as WithTypeName<Item>
    addItemToFieldList({
      cache,
      target: Parent,
      field: 'Children',
      item: result.createItem,
    })
  }
}
```

and finally a hook:

```typescript
export const useCreateItem = (
  options?: CreateItemOptions
): [
  HandleMethod<CreateItemData>,
  UseMutationState<CreateItemData, ItemVariables>
] => {
  const [{data, ...restResult}, mutate] = useMutation<
    CreateItemData,
    ItemVariables
  >(CREATE_ITEM, {
    ...options,
  })

  useEffect(() => {
    if (data) {
      // *** data.createItem will be null here
    }
  }, [JSON.stringify(data)])

  return [mutate, {data, ...restResult}]
```

When the mutation is complete, `data.createItem` in the `useEffect` is empty. However, when not setting `item.Parent`, `data.createItem` returns. In both cases, I can see the data returned successfully in the Chrome network tab.

Thanks in advance.

**Comment by kitten** [comment, marked as answer]
I think it may be related to your list containing `undefined`. Those aren't filtered out in the updater and will explicitly be stored as such and indicate that data is missing. I assume the intention there is to not spread the array at all (or spread an empty array / undefined) rather than the array containing undefined?

**Reply by orbiteleven** [reply]
Ah, yes. That was my bad. Changing my list generation to:
```typescript
const list = [
        ...(prepend ? [Item] : []),
        ...(cache.resolve(Target, fieldKey) as NullArray<string>),
        ...(prepend ? [] : [Item]),
      ]
```
Indeed solves the issue.

Followup question: what value is there in not automatically filtering out null/undefined entries? Just a "separation of concerns" thing?

**Reply by kitten** [reply]
There are cases where `undefined` is legitimately useful for Graphcache, since it's not represented in the spec and `null` is a valid resolution value.

Internally, during traversal that's all that's needed for us to recognise that a value is missing, to simplify the current logic a little. Sometimes it makes sense to map and replace them, sometimes it doesn't. In terms of updaters and resolvers, when you resolve values manually, you may still want to indicate missing values without adding a lot of checks on your own and cascade the main return value to `undefined`, hence we've decided to not hide `undefined` values at all

**Reply by orbiteleven** [reply]
One place this also comes up is using something like:
```
cache.invalidate(item)
```
It seems like Graphcache should be smart enough to not leave a bunch of empty holes.

**Reply by orbiteleven** [reply]
Oops, just saw your response, @kitten. Yeah, I guess that makes sense. I'd be curious to hear the 80/20 on both use cases, but understand y'all's decision on it.

**Reply by kitten** [reply]
When you invalidate data there isn't enough information to fill in the blanks. In other words, the presence of `null` isn't the same as the absence of data. That's why we have schema awareness, because GraphQL can actually help us get hints at where absences of data is acceptable, via what's nullable and what isn't. A spec proposal is actually aiming to take that concept further into queries: https://github.com/FormidableLabs/urql/pull/2018
